File: library/src/main/java/name/zeno/android/util/Rx2Timer.java
package name.zeno.android.util;

import android.content.Intent;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;

import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import lombok.Setter;
import name.zeno.android.exception.ZException;
import name.zeno.android.listener.Action0;
import name.zeno.android.listener.Action1;
import name.zeno.android.presenter.LifecycleListener;
import name.zeno.android.presenter.ZFragment;
import name.zeno.android.third.rxjava.ZObserver;

/**
 * @author 陈治谋 (513500085@qq.com)
 * @since 2017/5/9
 */
public class Rx2Timer implements LifecycleListener
{
  private static final String TAG = "Rx2Timer";
  private Disposable disposable;
  @Setter private long expiredIn = 0;
  // 周期 s
  @Setter
  private         int  period    = 1;
  @Setter private Action1<Long> onNext;
  @Setter private Action0       onComplete;


  private Rx2Timer() { }

  public void start()
  {
    if (period < 1) period = 1;
    Observable.interval(0, period, TimeUnit.SECONDS)
        .takeWhile(value -> System.currentTimeMillis() <= expiredIn + 1000)
        .subscribeOn(Schedulers.single())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new ZObserver<Long>()
        {
          @Override public void onNext(Long value)
          {
            if (onNext != null) {
              long remain = (expiredIn - System.currentTimeMillis()) / 1000;
              onNext.call(remain < 0 ? 0 : remain);
            }
          }

          @Override public void onComplete()
          {
            if (onComplete != null) {onComplete.call();}
          }

          @Override public void handleError(ZException e)
          {
            ZLog.e(e.getMessage());
          }

          @Override public void onSubscribe(Disposable d)
          {
            super.onSubscribe(d);
            disposable = d;
          }
        });
  }

  public void stop()
  {
    if (disposable != null && !disposable.isDisposed()) {
      disposable.dispose();
      disposable = null;
    }
    expiredIn = 0;
  }

  public static Rx2Timer with(FragmentActivity activity)
  {
    Rx2Timer rx2Timer = new Rx2Timer();
    rx2Timer.registLifecycle(activity.getSupportFragmentManager());
    return rx2Timer;
  }

  public static Rx2Timer with(Fragment fragment)
  {
    Rx2Timer rx2Timer = new Rx2Timer();
    rx2Timer.registLifecycle(fragment.getFragmentManager());
    return rx2Timer;
  }

  //<editor-fold desc="lifecycle listener">
  @Override public void onCreate() { }

  @Override public void onResume() { }

  @Override public void onActivityResult(int requestCode, int resultCode, Intent data) { }

  @Override public void onViewCreated() { }

  @Override public void onDestroyView() { }

  @Override public void onDestroy()
  {
    stop();
  }
  //</editor-fold>

  private void registLifecycle(FragmentManager fragmentManager)
  {
    ZFragment fragment = (ZFragment) fragmentManager.findFragmentByTag(TAG);
    if (fragment == null) {
      fragment = new ZFragment();
      fragmentManager.beginTransaction().add(fragment, TAG).commit();
    }
    fragment.registerLifecycleListener(this);
  }
}

File: library/src/main/java/name/zeno/android/util/Rx2Timer.kt
package name.zeno.android.util

import android.content.Intent
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.support.v4.app.FragmentManager

import java.util.concurrent.TimeUnit

import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import name.zeno.android.exception.ZException
import name.zeno.android.listener.Action0
import name.zeno.android.listener.Action1
import name.zeno.android.presenter.LifecycleListener
import name.zeno.android.presenter.ZFragment
import name.zeno.android.third.rxjava.ZObserver

/**
 * @author 陈治谋 (513500085@qq.com)
 * @since 2017/5/9
 */
class Rx2Timer private constructor() : LifecycleListener {
  private var disposable: Disposable? = null
  private var expiredIn: Long = 0
  // 周期 s
  private var period = 1
  private var onNext: Action1<Long>? = null
  private var onComplete: Action0? = null

  fun start() {
    if (period < 1) period = 1
    Observable.interval(0, period.toLong(), TimeUnit.SECONDS)
        .takeWhile { value -> System.currentTimeMillis() <= expiredIn + 1000 }
        .subscribeOn(Schedulers.single())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(object : ZObserver<Long>() {
          override fun onNext(value: Long) {
            if (onNext != null) {
              val remain = (expiredIn - System.currentTimeMillis()) / 1000
              onNext!!.call(if (remain < 0) 0 else remain)
            }
          }

          override fun onComplete() {
            if (onComplete != null) {
              onComplete!!.call()
            }
          }

          override fun handleError(e: ZException) {
            ZLog.e(e.message ?: "")
          }

          override fun onSubscribe(d: Disposable) {
            super.onSubscribe(d)
            disposable = d
          }
        })
  }

  fun stop() {
    if (disposable != null && !disposable!!.isDisposed) {
      disposable!!.dispose()
      disposable = null
    }
    expiredIn = 0
  }

  //<editor-fold desc="lifecycle listener">
  override fun onCreate() {}

  override fun onResume() {}

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {}

  override fun onViewCreated() {}

  override fun onDestroyView() {}

  override fun onDestroy() {
    stop()
  }
  //</editor-fold>

  private fun registLifecycle(fragmentManager: FragmentManager) {
    var fragment: ZFragment? = fragmentManager.findFragmentByTag(TAG) as ZFragment
    if (fragment == null) {
      fragment = ZFragment()
      fragmentManager.beginTransaction().add(fragment, TAG).commit()
    }
    fragment.registerLifecycleListener(this)
  }

  fun setExpiredIn(expiredIn: Long) {
    this.expiredIn = expiredIn
  }

  fun setPeriod(period: Int) {
    this.period = period
  }

  fun setOnNext(onNext: Action1<Long>) {
    this.onNext = onNext
  }

  fun setOnComplete(onComplete: Action0) {
    this.onComplete = onComplete
  }

  companion object {
    private val TAG = "Rx2Timer"

    fun with(activity: FragmentActivity): Rx2Timer {
      val rx2Timer = Rx2Timer()
      rx2Timer.registLifecycle(activity.supportFragmentManager)
      return rx2Timer
    }

    fun with(fragment: Fragment): Rx2Timer {
      val rx2Timer = Rx2Timer()
      rx2Timer.registLifecycle(fragment.fragmentManager)
      return rx2Timer
    }
  }
}
